File: src/tuni_whitegoods_projector/scripts/tuni_whitegoods_projector/button.py
import cv2


class Button():
    def __init__(self, id_, zone, name, description, text, b_color, t_color, center, radius, hidden):
        self._id = id_
        self._zone = zone
        self._name = name
        self._description = description
        self._text = text
        self._button_color = (b_color.b*255, b_color.g*255, b_color.r*255)
        self._text_color = (t_color.b*255, t_color.g*255, t_color.r*255)
        self._center = (center[0], center[1])
        self._radius = int(radius)
        self._hidden = hidden

    def set_button_color(self,b_color):
        self._button_color = (b_color.b*255, b_color.g*255, b_color.r*255)

    def get_center(self):
        return self._center
    
    def get_zone(self):
        return self._zone
    
    def draw_button(self, img):
        if not self._hidden:
            img = cv2.circle(img, self._center, self._radius, self._button_color, -1)
            TEXT_FACE = cv2.FONT_HERSHEY_DUPLEX
            TEXT_SCALE, text_size, TEXT_THICKNESS = self.get_text_attributes()
            text_origin = (self._center[0] - text_size[0] // 2, self._center[1] + text_size[1] // 2)
            cv2.putText(img, self._text, text_origin, TEXT_FACE, TEXT_SCALE, self._text_color, TEXT_THICKNESS, cv2.LINE_AA)

        return img
    
    def get_text_attributes(self):
        TEXT_FACE = cv2.FONT_HERSHEY_DUPLEX
        TEXT_SCALE = 0.1
        TEXT_THICKNESS = 2
        right_size = False
        text_size, _ = cv2.getTextSize(self._text, TEXT_FACE, TEXT_SCALE, TEXT_THICKNESS)
        if text_size[0] > 60:
            TEXT_THICKNESS = 1
        while not right_size:
            if (text_size[0] < self._radius*2-10 and text_size[0] > self._radius*2-25): 
                right_size = True
            else:
                TEXT_SCALE = TEXT_SCALE + 0.1 
            if TEXT_SCALE > 10:
                right_size = True
            text_size, _ = cv2.getTextSize(self._text, TEXT_FACE, TEXT_SCALE, TEXT_THICKNESS)

        return TEXT_SCALE, text_size, TEXT_THICKNESS
    
    def print_button(self):
        print(self._id)
        print(self._zone)
        print(self._button_color)

File: tuni_whitegoods_projector_interface/src/tuni_whitegoods_projector_interface/button.cpp
#include "tuni_whitegoods_projector_interface/button.h"

Button::Button(std::string request_id, std::string name, std::string text, 
               std_msgs::ColorRGBA button_color, std_msgs::ColorRGBA text_color, 
               geometry_msgs::Pose center, float radius) 
			: name(name),
		    text(text),
			  center(center),
			  radius(radius),
        id(request_id)
{
  set_button_color(button_color);
  set_text_color(text_color);
  button_already_pressed = false;
}

cv::Mat Button::draw() {
  btn_img = cv::Mat::zeros(1080, 1920, CV_8UC3);

  cv::circle(btn_img, center_cam_point, radius, btn_color, -1);

  int fontFace = cv::FONT_HERSHEY_SIMPLEX;
  double fontScale = 0.5;
  int thickness = 2;

  // Calculate the size of the text box
  int baseline = 0;
  cv::Size textSize = cv::getTextSize(text, fontFace, fontScale, thickness, &baseline);

  // Calculate the bottom-left corner of the text box to center it
  cv::Point text_origin(
      center_cam_point.x - textSize.width / 2,
      center_cam_point.y + textSize.height / 2
  );

  cv::putText(btn_img, text, text_origin, fontFace, fontScale, txt_color, thickness);

  return btn_img;
}

bool Button::isAlreadyPressed(){
  return button_already_pressed;
}

void Button::setAlreadyPressed(bool alreadyPressed){
  button_already_pressed = alreadyPressed;
}

bool Button::checkForInteractions(std::string name,
                                        cv::Point hand_position) {
  bool result = false;
  float distance = cv::norm(hand_position - center_cam_point);
  bool is_crossed = distance < radius*1.2;
  if (name == "left") {
    left_hand_press = is_crossed;
  } else if (name == "right") {
    right_hand_press = is_crossed;
  }

  button_pressed = (left_hand_press || right_hand_press);

  if (button_pressed){
    btn_color = cv::Scalar(0, 0, 255);
    result = true;
  } else {
    btn_color = cv::Scalar(255,0, 0);  }  
  return result;
}

void Button::resetInteractions() {
  button_pressed = false;
  btn_color = cv::Scalar(255, 0, 0);
}

std::string Button::get_name() {
  return name;
}

void Button::set_button_color(std_msgs::ColorRGBA button_color) {
  ros_btn_color = button_color;
  base_btn_color= cv::Scalar(button_color.b * 255, button_color.g * 255, button_color.r * 255, button_color.a * 255);
  btn_color = cv::Scalar(button_color.b * 255, button_color.g * 255, button_color.r * 255, button_color.a * 255);
}

void Button::set_text_color(std_msgs::ColorRGBA text_color) {
  ros_text_color = text_color;
  txt_color = cv::Scalar(text_color.b * 255, text_color.g * 255, text_color.r * 255, text_color.a * 255);
}

std::string Button::getId() { return id; }


            